# 01. Importing Libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02. Importing Data

In [2]:
# Project folder pathway

path = r'C:\Users\SD.DESKTOP-45BHU58\OneDrive\Desktop\CF\Section 4 - Python\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\SD.DESKTOP-45BHU58\\OneDrive\\Desktop\\CF\\Section 4 - Python\\Instacart Basket Analysis'

In [4]:
# Importing updated ords_prods_merged

ords_prods_merged = pd.read_pickle(r'C:\Users\SD.DESKTOP-45BHU58\OneDrive\Desktop\CF\Section 4 - Python\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged_3.pkl')

In [5]:
# Importing customers data set

df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col=False)

In [6]:
# Checking shape

df_cust.shape

(206209, 10)

In [7]:
# Checking columns

df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# 03. Data Wrangling

### Renaming Columns

In [8]:
# Changing First Name to first_name

df_cust.rename(columns = {'First Name': 'first_name'}, inplace = True)

In [9]:
# Changing Surnam to last_name

df_cust.rename(columns = {'Surnam': 'last_name'}, inplace = True)

In [10]:
# Changing Gender to gender

df_cust.rename(columns = {'Gender': 'gender'}, inplace = True)

In [11]:
# Changing STATE to state

df_cust.rename(columns = {'STATE': 'state'}, inplace = True)

In [12]:
# Changing Age to age

df_cust.rename(columns = {'Age': 'age'}, inplace = True)

In [13]:
# Changing n_dependants to number_of_dependants

df_cust.rename(columns = {'n_dependants': 'number_of_dependants'}, inplace = True)

In [14]:
# Double checking renamed columns

df_cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants fam_status  income  
0                     3    married  165665  
1                     0     single   59285  
2                     2    married   99568  
3                     0     single   42049  
4                     1    married   40374

### Changing Data Type

In [15]:
# Descriptive analysis

df_cust.describe()

user_id            age  number_of_dependants         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [16]:
# Changing user_id from int to str - statistics won't need to be calculated for this column

df_cust['user_id'] = df_cust['user_id'].astype('str')

In [17]:
# Double checking data type for 'user_id'

df_cust['user_id'].dtype

dtype('O')

In [18]:
# Checking ALL data types

df_cust.dtypes

user_id                 object
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
fam_status              object
income                   int64
dtype: object

### No columns will be dropped nor will this dataframe need to be transposed

# 04. Data Consistency Checks

In [19]:
# Checking data accuracy

df_cust.describe()

age  number_of_dependants         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

Data overall looks accurate. The income max of 593,901 looks odd but it is possible for an individual with a high earning job to make this amount.

### Missing Values

In [20]:
# Listing columns and count of missing values

df_cust.isnull().sum()

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
fam_status                  0
income                      0
dtype: int64

In [21]:
# Checking rows with 'first_name' columns containing null values

df_cust[df_cust['first_name'].isnull() == True]

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependants fam_status  income  
53                         2    married   41709  
73                         0     single   82518  
82                         3    married  117099  
99                         1    married  155673  
105                        3    married  151819  
...                      ...        ...     ...  
206038                     3    married   87783  
206044                     2    married  108488  
206090                     1    married   45275  
206162                     0     single   56325  
206171                     2    married   59222  

[11259 rows x 10 columns]

I will not remove records containing missing values under 'first_name' column since it still contains valuable information

### Mixed Data Types

In [22]:
# Checking for columns with mixed data types

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

first_name


In [23]:
# Changing 'first_name' column to string

df_cust['first_name'] = df_cust['first_name'].astype('str')

In [24]:
# Double checking for columns with mixed data types

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

No columns printed out on second check

### Duplicates

In [25]:
# Checking for duplicated rows

df_cust[df_cust.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, fam_status, income]
Index: []

No duplicate rows found

# 05. Exporting Wrangled/Checked Customer Data Set

In [26]:
# Exporting file as pickle

df_cust.to_pickle(os.path.join(path,'02 Data','Prepared Data','customers_checked.pkl'))

# 06. Combining Data Sets

In [27]:
# Importing updated customers dataframe

df_custs = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','customers_checked.pkl'))

In [28]:
# Double checking shape of df_custs

df_custs.shape

(206209, 10)

In [29]:
# Double checking headers of df_custs

df_custs.head()

user_id first_name last_name  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants fam_status  income  
0                     3    married  165665  
1                     0     single   59285  
2                     2    married   99568  
3                     0     single   42049  
4                     1    married   40374

In [30]:
# Double checking shape of ords_prods_merged

ords_prods_merged.shape

(32404859, 25)

In [31]:
# Double checking headers of ords_prods_merged

ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ...    price_range_loc     busiest_day         busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly busy days   
1          1  ...  Mid-range product  Regularly busy      Least busy days   
2          1  ...  Mid-range product  Regularly busy      Least busy days   
3          1  ...  Mid-range product      Least busy      Least busy days   
4          1  ...  Mid-range product      Least busy      Least busy days   

   busiest_period_of_day max_order  loyalty_flag avg_order_cost_by_cust  \
0         Average orders        10  New customer               6.367797   
1         Average orders        10  New customer               6.367797   
2            Most orders        10  New customer               6.367797   
3         Average orders        10  New customer               6.367797   
4            Most orders        10  New customer               6.367797   

  spender_flag order_frequency_by_cust         frequency_flag  
0  Low spender                    20.5  Non-frequent customer  
1  Low spender                    20.5  Non-frequent customer  
2  Low spender                    20.5  Non-frequent customer  
3  Low spender                    20.5  Non-frequent customer  
4  Low spender                    20.5  Non-frequent customer  

[5 rows x 25 columns]

In [32]:
# Checking data type of df_custs 'user_id'

df_custs['user_id'].dtype

dtype('O')

In [33]:
# Checking data type of ords_prods_merged 'user_id'

ords_prods_merged['user_id'].dtype

dtype('int64')

In [34]:
# Converting ords_prods_merged 'user_id' to str - this way 'user_id' from both data sets are the same data type

ords_prods_merged['user_id'] = ords_prods_merged['user_id'].astype('str')

In [35]:
# Double checking ords_prods_merged 'user_id' data type

ords_prods_merged['user_id'].dtype

dtype('O')

In [36]:
# Merging data sets

ords_prods_cust_merged = ords_prods_merged.merge(df_custs, on = 'user_id')

In [37]:
# Double checking merged shape

ords_prods_cust_merged.shape

(32404859, 34)

In [38]:
# Checking all columns

pd.options.display.max_columns = None
ords_prods_cust_merged.head()

order_id user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329       1             1                       2                  8   
1   2398795       1             2                       3                  7   
2    473747       1             3                       3                 12   
3   2254736       1             4                       4                  7   
4    431534       1             5                       4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc     busiest_day         busiest_days  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product  Regularly busy      Least busy days   
2  Mid-range product  Regularly busy      Least busy days   
3  Mid-range product      Least busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  max_order  loyalty_flag  avg_order_cost_by_cust  \
0        Average orders         10  New customer                6.367797   
1        Average orders         10  New customer                6.367797   
2           Most orders         10  New customer                6.367797   
3        Average orders         10  New customer                6.367797   
4           Most orders         10  New customer                6.367797   

  spender_flag  order_frequency_by_cust         frequency_flag first_name  \
0  Low spender                     20.5  Non-frequent customer      Linda   
1  Low spender                     20.5  Non-frequent customer      Linda   
2  Low spender                     20.5  Non-frequent customer      Linda   
3  Low spender                     20.5  Non-frequent customer      Linda   
4  Low spender                     20.5  Non-frequent customer      Linda   

  last_name  gender    state  age date_joined  number_of_dependants  \
0    Nguyen  Female  Alabama   31   2/17/2019                     3   
1    Nguyen  Female  Alabama   31   2/17/2019                     3   
2    Nguyen  Female  Alabama   31   2/17/2019                     3   
3    Nguyen  Female  Alabama   31   2/17/2019                     3   
4    Nguyen  Female  Alabama   31   2/17/2019                     3   

  fam_status  income  
0    married   40423  
1    married   40423  
2    married   40423  
3    married   40423  
4    married   40423

# 07. Making Changes from Practice Example

Changes from practice example were not saved. Re-doing the step so it does not affect visualizations for Part 2.

In [41]:
# Check stats for 'prices' column

ords_prods_cust_merged['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [42]:
#Set values >100 as NaN / missing

ords_prods_cust_merged.loc[ords_prods_cust_merged['prices'] >100, 'prices'] = np.nan

In [43]:
# Double checked 'prices' stats

ords_prods_cust_merged['prices'].describe()

count    3.239973e+07
mean     7.790994e+00
std      4.241809e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

# 08. Exporting Combined Data

In [45]:
# Export ords_prods_cust_merged as pickle file

ords_prods_cust_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))